test/monitoring/monitoring_test.go

```diff
@@ -0,0 +1,69 @@
+// Copyright 2020 The Lokomotive Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build aws packet
+// +build poste2e
+
+package monitoring
+
+import (
+	"fmt"
+	"testing"
+
+	testutil "github.com/kinvolk/lokomotive/test/components/util"
+
+	"github.com/prometheus/client_golang/api"
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
+)
+
+func TestPrometheus(t *testing.T) {
```

Comment:
Hm, I think that reduces test readability/makes it more complex. It's fine to even have repetition in tests, they should be as straightforward as possible.

That also seems like a premature optimization, if we have no other tests to add at the moment.

If we're going to have more tests, I would suggest alternative approach. How about adding `getPrometheusClient()` function, which will return prometheus client? Then each test will just call this function at the beginning to get a client instance and then they can proceed on their own. I'm aware that this will create multiple instances of the client, but I don't think that should be a problem.

Reply:
> That also seems like a premature optimization, if we have no other tests to add at the moment.

This is not premature optimization because it will be consumed in https://github.com/kinvolk/lokomotive/pull/120 and https://github.com/kinvolk/lokomotive/pull/140

>  I would suggest alternative approach. How about adding getPrometheusClient() function, 

The problem with this approach is that then who takes care of closing the ports open for port forwarding? The whole point of the PR is to reuse the prometheus client. Since every port forward request consumes almost 3-4 seconds and for every test we cannot afford to have one. This reuses the client connection and the same function also takes care of closing the connection.

Reply:
>This is not premature optimization because it will be consumed in #120 and #140

That has not been mentioned in original PR message. It make sense then.

>The problem with this approach is that then who takes care of closing the ports open for port forwarding? The whole point of the PR is to reuse the prometheus client. Since every port forward request consumes almost 3-4 seconds and for every test we cannot afford to have one. This reuses the client connection and the same function also takes care of closing the connection.


That make sense then, but again, it has not been mentioned in the commit message nor in the PR message. Can we then please both tests in the single file, so they are easier to read? I would even consider flattening the test matrix, as nesting in twice seems really odd to me.

Reply:
Also, please document the reasoning of chosen approach in some comments to test test, so it's not confusing for other people too.

Reply:
> Can we then please both tests in the single file, so they are easier to read?

I think keeping it in separate files is more readable IMHO because then it reduces the endless scroll. Also with growing number of things we add to the test the files ought to increase in length.

> I would even consider flattening the test matrix, as nesting in twice seems really odd to me.

I can flatten it yes.

Reply:
>I think keeping it in separate files is more readable IMHO because then it reduces the endless scroll. Also with growing number of things we add to the test the files ought to increase in length.

The problem with 2 files is, that in order to understand the test, you need to open 2 files. As far as I remember, that should be avoided for simplicity, if things are related to each other. I wonder what other think about it.

Reply:
Things are not same, because in one we are testing what all metrics are loaded and in other we are seeing the scrape targets and in other we are looking at alerts.